docs/design/DAM-on-type-and-RUC-interaction.md

```diff
@@ -0,0 +1,151 @@
+
+## RUC or DAM on nested type
+
+```csharp
+[DAM(All)]
+class AnnotatedBase {}
+
+class Derived : AnnotatedBase
+{
+    [RUC]
+    class NestedWithRUC {}
+
+    [DAM]
+    class NestedWithDAM {}
+}
+```
+
+This is probably an existing hole. Accessing a RUC or DAM type itself via reflection is not problematic on its own (since both annotations should behave as applying to members only). But it can create a hole in the analysis because of annotation propagation through `GetNestedType`. For example (given the above code):
+
+```csharp
+void TestMethod(Derived instance)
+{
+    Type typeOfDerived = instance.GetType(); // The type value is effectively annotated with All
+    Type nestedWithRUC = typeOfDerived.GetNestedType("NestedWithRUC"); // The resulting type value is effectively annotated with All
+    Activator.CreateInstance(nestedWithRUC); // No warning - nestedWithRUC has All annotation on it
+}
+```
+
+The above MUST warn, since it's calling a `.ctor` on a RUC annotated type.
+
+DAM on nested type doesn't seem to have a problem mainly because for marking purposes it's not needed (the DAM on the outer type if it applies to the nested type will mark the entire nested type, so DAM on it can't mark more).
+Also accessing DAM annotated types themselves is not a problem, and accessing their members is drive by the DAM annotation.
+This needs verification!!! (It is tricky)
+
+## Conclusion
+
+Personally I'm not a big fan of the fact that RUC on type doesn't suppress the new warnings on the type's members - as a normal user I would expect that to happen. That said, doing so would introduce an "Exception" to the existing rules, which is not ideal either. If we warn more (RUC on type doesn't affect the new warnings), it sort of creates noise, but it's not exactly wrong. And we can change our mind later on (by not warning anymore some day in the future).
+
+In the end the interaction is very simple, the new warnings caused by DAM on type are not affected by RUC in any way (as in, they're never suppressed by RUC).
+
+## Samples
+
+```csharp
+[DAM(All)]
+class AnnotatedBase {}
+
+[RUC]
+class DerivedWithRUC : AnnotatedBase
+{
+    // IL2112
+    [RUC]
+    public void MethodWithRUC() {}
+
+    // IL2114
+    public static void MethodWithDAM([DAM] Type type) {}
+
+    // IL2112
+    // IL2114
+    [RUC]
+    public void MethodWithDAMAndRUC([DAM] Type type) {}
+
+    // IL2114
+    [DAM]
+    public Type FieldWithDAM;
+
+    // IL2112 - we must do this here, since we won't be able to see the RUC anywhere else
+    [RUC]
+    public class NestedWithRUC {}
+
+    // No warnings - ??? Probably ??? - I think the DAM on the outer type effectively implies DAM(All) on the nested type
+    // and thus declaring it explicitly has no effect. But this needs validation specifically for cases where the outer type
+    // has DAM(PublicNestedTypes) on it (the new behavior should still imply All on the nested type in such case I think)
```

Comment:
Agreed, a DAM nested type shouldn't warn because of the outer DAM.

A related question is whether RUC members of the nested DAM type should produce two warnings - one for the outer DAM, one for the inner DAM. I think it should - while the outer DAM effectively implies the inner DAM, it's possible that the app only uses the nested type's DAM requirements (they are really separate references to the RUC member).

Reply:
Agreed - I would not worry about producing "duplicate" warnings at this point - that's something we can worry about later if needed.